text/3724-theupdateframework.md

```diff
@@ -0,0 +1,341 @@
+- An update to the tuf-crates-root.json file will occur in the `tuf-root` repository, which shall trigger a new singing event Pull Request, which the root quorum must perform.
+
+## (cargo-tuf-lib) Standard TUF Implementation
+
+We propose creating a new crate, `cargo-tuf-lib`, which shall be used by both Cargo and Rustup for doing TUF synchronization and update procedures. This library shall be a shim wrapper around the `rust-tuf` crate (https://github.com/rustfoundation/rust-tuf), providing a simplified and shared interface for doing synchronization and verification of the TUF repositories and their files. 
+
+The API surface of this crate is to be determined upon implementation in Cargo and Rustup. However, because both tools will need to perform synchronization and validation against the tuf-root repository, they shall used this shared interface to guarantee compatibility.
+
+This API will include operations to sync the TUF repositories efficiently, and to perform a verified download of an object.
+
+## TAP-16 Implementation
+
+We're proposing to use [TAP-16](https://github.com/theupdateframework/taps/blob/master/tap16.md) to provide efficient update checking and download sizes. TAP-16 uses Merkle trees rather than full lists for the download of a snapshot of the inventory of a repository (`snapshot.json`). We want to ensure that, as crates.io grows, the total size clients have to download when checking for updates remains small.
+
+### Shared local folder changes
+
+Creation of a new `~/.cargo/tuf` directory. (If Cargo stores its registry information in another directory, the `tuf` directory should be stored alongside the `registry` directory.) This directory shall be used for all TUF operations by project tools (both Rustup and Cargo). The cargo folder was chosen as the main location of residence for these files given that although Rustup will be performing the initialization of these folders, there is already a precedent set for shared files living within the cargo folder.
+
+- `~/.cargo/tuf` The top-level directory of local copies of TUF repositories
+- `~/.cargo/tuf/root` a copy of the `tuf-root` repository locally synchronized
+- `~/.cargo/tuf/crates` a copy of the `tuf-crates` repository locally synchronized
+
+### Controlling enablement of TUF
+
+This RFC does not specify how to handle non-`crates.io` repositories. Cargo can choose to enable TUF for third-party repositories in the future, or may default to only using TUF for crates.io unless otherwise configured. Cargo might choose to use an environment variable (e.g. `CARGO_TUF_DISABLE`) to disable all usage of TUF on third-party repositories.
+
+## Cargo changes
+- Addition of `cargo-tuf-lib` as a dependency
+- `cargo-tuf-lib::sync` attempted prior to an index update
+- `cargo-tuf-lib::verify_snapshot` called on an index update on the entire index
```

Comment:
With the sparse registry, we don't have a "index update" phase but we update the parts of the registry if-needed as we perform a registry operation.  We also don't have the entire index.

Do we need to do this sync even if we won't download anything new from the registry?    Could we instead only check if there was a change?  Can we only check what changed or what is downloaded?